### PR TITLE
Bump tectonic to 0.3.3

### PR DESCRIPTION
**Pull request by MichaelHatherly**
From what I can see locally, the `rust` version the BB provides is now too old to build this. What's needed to update that?

### CODE REVIEW

**Comment by giordano** [comment]
> What's needed to update that?

Carve some time out of @staticfloat to have a look at it, good luck with that :stuck_out_tongue:

**Comment by giordano** [comment]
https://github.com/JuliaPackaging/Yggdrasil/releases/tag/RustToolchain-v1.48.0 :eyes:

**Comment by staticfloat** [comment]
Once https://github.com/JuliaPackaging/Yggdrasil/pull/2152 gets merged, rebase this on top of that and try again.  :)

**Comment by MichaelHatherly** [comment]
Thanks for the rapid response @staticfloat. I'll give it a go again once that one's merged. (In the meantime I'll just go with bundling the few official binaries the tectonic provides themselves, so no rush needed from my side.)

**Comment by giordano** [comment]
All failures look the same:
```
[14:22:30] error: linker `x86_64-apple-darwin-gcc` not found
[14:22:30]   |
[14:22:30]   = note: No such file or directory (os error 2)
```
@staticfloat any clue?

**Comment by staticfloat** [comment]
There appears to be a mismatch in the `os_version` stuff; I think `rust` is looking for `x86_64-apple-darwin-gcc` when it should be looking for `x86_64-apple-darwin14-gcc`.

**Comment by staticfloat** [comment]
I think we just need to change `CARGO_BUILD_TARGET` to be `x86_64-apple-darwin14` instead of `x86_64-apple-darwin`.

**Comment by giordano** [comment]
Uhm, no, if I do that I get
```
sandbox:${WORKSPACE}/srcdir/tectonic-tectonic-0.3.3 # cargo build --release -j${nproc}
error: failed to run `rustc` to learn about target-specific information

Caused by:
  process didn't exit successfully: `rustc - --crate-name ___ --print=file-names --target x86_64-unknown-freebsd11.1 --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=cfg` (exit code: 1)
  --- stderr
  error: Error loading target specification: Could not find specification for target "x86_64-unknown-freebsd11.1". Use `--print target-list` for a list of built-in targets
```
and
```
sandbox:${WORKSPACE}/srcdir/tectonic-tectonic-0.3.3 # rustc --print target-list | grep -E '(freebsd|armv7|darwin)'
aarch64-apple-darwin
aarch64-unknown-freebsd
armv6-unknown-freebsd
armv7-apple-ios
armv7-linux-androideabi
armv7-unknown-cloudabi-eabihf
armv7-unknown-freebsd
armv7-unknown-linux-gnueabi
armv7-unknown-linux-gnueabihf
armv7-unknown-linux-musleabi
armv7-unknown-linux-musleabihf
armv7-unknown-netbsd-eabihf
armv7-wrs-vxworks-eabihf
armv7a-none-eabi
armv7a-none-eabihf
armv7r-none-eabi
armv7r-none-eabihf
armv7s-apple-ios
i686-apple-darwin
i686-unknown-freebsd
powerpc64-unknown-freebsd
x86_64-apple-darwin
x86_64-unknown-freebsd
```
The OS version doesn't seem to be the issue (and wouldn't explain the error for armv7l)

**Comment by staticfloat** [comment]
But doesn't your error look like it passed through on darwin and instead went through to freebsd, where we need to do the same thing?

**Comment by giordano** [comment]
I ran it on FreeBSD :grin:

**Comment by staticfloat** [comment]
Ah, but the error now is that rust doesn't understand that triplet with the OS version.  Sorry, I got sidetracked by the switch from darwin to freebsd.  Hmmm.  What if we generate symlinks that go from `x86_64-apple-darwin-gcc -> x86_64-apple-darwin14-gcc`.  Does that work?

**Comment by giordano** [comment]
The error on armv7l is
```
[14:21:28] error: linker `armv7l-linux-gnueabihf-gcc` not found
[14:21:28]   |
[14:21:28]   = note: No such file or directory (os error 2)
[14:21:28] 
[14:21:28] error: aborting due to previous error
```
which suggests that the platform is inferred from something else, `CARGO_BUILD_TARGET` has armv7l -> armv7

**Comment by giordano** [comment]
Oh, and for armv7l the linker should be `arm-linux-gnueabihf-gcc`, not `armv7l-linux-gnueabihf-gcc` or `armv7-linux-gnueabihf-gcc`

**Comment by giordano** [comment]
Ok, I can confirm that by manually fixing `/opt/${MACHTYPE}/config` I can successfully compile for armv7l :tada:

**Comment by giordano** [comment]
https://github.com/JuliaPackaging/BinaryBuilderBase.jl/pull/80 should work around all issues, https://github.com/JuliaPackaging/Yggdrasil/pull/2168 fixes the configuration using `armv*l` instead of `arm`

**Comment by giordano** [comment]
```
[00:49:13] error: cannot produce proc-macro for `failure_derive v0.1.8` as the target `x86_64-unknown-linux-musl` does not support these crate types
[00:49:13]  ---> cargo build --release -j${nproc}
```
:disappointed:

**Comment by giordano** [comment]
Try https://github.com/rust-lang/cargo/issues/7154#issuecomment-561947609
```
RUSTFLAGS="-C target-feature=-crt-static"
```
(not sure where though)?

**Comment by giordano** [comment]
It didn't help much :confused:

File: T/tectonic/build_tarballs.jl [hunk 1]
```diff
@@ -49,7 +52,7 @@ dependencies = [
     Dependency("FreeType2_jll"),
     Dependency("Graphite2_jll"),
     Dependency("HarfBuzz_jll"),
```
Comment (giordano) [inline T/tectonic/build_tarballs.jl:54]:
```suggestion
    Dependency("HarfBuzz_jll", v"2.8.1"; compat="2.8.1"),
    Dependency("HarfBuzz_ICU_jll", v"2.8.1"; compat="2.8.1"),
```

File: T/tectonic/build_tarballs.jl [hunk 2]
```diff
@@ -49,7 +52,7 @@ dependencies = [
...
-    Dependency(PackageSpec(; name="ICU_jll", version=v"67.1.0")),
+    Dependency("ICU_jll", v"67.1.0", compat="67.1.0"),
```
Comment (giordano) [inline T/tectonic/build_tarballs.jl:55]:
```suggestion
    Dependency("ICU_jll", v"69.1.0"; compat="69.1.0"),
```

**Review by giordano — COMMENTED** [review]
If you are ok to require Julia v1.6, we should do this (after having a working Rust toolchain...)

**Comment by MichaelHatherly** [comment]
> If you are ok to require Julia v1.6

Fine by me :+1:

**Comment by MichaelHatherly** [comment]
(Current latest tectonic version is now `0.4.1` rather than `0.3.3` as is referenced in this PR.)

**Pull request closed without merging**